presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -15,283 +15,367 @@
+            assignedSplits.set(i, null);
         }
 
-        private PrestoSparkTaskDescriptor createTaskDescriptor(PlanFragment fragment, List<ScheduledSplit> splits)
-        {
-            Map<PlanNodeId, Set<ScheduledSplit>> splitsByPlanNode = splits.stream()
-                    .collect(Collectors.groupingBy(
-                            ScheduledSplit::getPlanNodeId,
-                            mapping(identity(), toSet())));
-
-            List<TaskSource> taskSourceByPlanNode = splitsByPlanNode.entrySet().stream()
-                    .map(entry -> new TaskSource(
-                            entry.getKey(),
-                            entry.getValue(),
-                            ImmutableSet.of(),
-                            true))
-                    .collect(toImmutableList());
-
-            return new PrestoSparkTaskDescriptor(
-                    session.toSessionRepresentation(),
-                    session.getIdentity().getExtraCredentials(),
-                    fragment,
-                    taskSourceByPlanNode,
-                    tableWriteInfo);
+        return sparkContext.parallelize(serializedTaskDescriptors.build(), numTasks)
+                .mapPartitionsToPair(createTaskProcessor(executorFactoryProvider, taskStatsCollector));
+    }
+
+    private List<ScheduledSplit> getSplits(Session session, TableScanNode tableScan)
+    {
+        List<ScheduledSplit> splits = new ArrayList<>();
+        SplitSource splitSource = splitManager.getSplits(session, tableScan.getTable(), UNGROUPED_SCHEDULING);
+        long sequenceId = 0;
+        while (!splitSource.isFinished()) {
+            List<Split> splitBatch = getFutureValue(splitSource.getNextBatch(NOT_PARTITIONED, Lifespan.taskWide(), 1000)).getSplits();
+            for (Split split : splitBatch) {
+                splits.add(new ScheduledSplit(sequenceId++, tableScan.getId(), split));
+            }
+        }
+        return splits;
+    }
+
+    private static List<List<ScheduledSplit>> assignSplitsToTasks(List<ScheduledSplit> splits, int numTasks)
+    {
+        checkArgument(numTasks > 0, "numTasks must be greater then zero");
+        List<List<ScheduledSplit>> assignedSplits = new ArrayList<>();
+        for (int i = 0; i < numTasks; i++) {
+            assignedSplits.add(new ArrayList<>());
+        }
+        for (int splitIndex = 0; splitIndex < splits.size(); splitIndex++) {
+            assignedSplits.get(splitIndex % numTasks).add(splits.get(splitIndex));
         }
+        return assignedSplits;
+    }
+
+    private PrestoSparkTaskDescriptor createIntermediateTaskDescriptor(Session session, TableWriteInfo tableWriteInfo, PlanFragment fragment)
+    {
+        return createSourceTaskDescriptor(session, tableWriteInfo, fragment, ImmutableList.of());
+    }
+
+    private PrestoSparkTaskDescriptor createSourceTaskDescriptor(
```

Comment:
When supporting bucket table join unbucketed table, does the join stage consider as `SourceTask`? -- it has to read from table, but also it's sort of a "intermediate task" since it has shuffle inputs.

Reply:
I haven't started prototyping bucketed table support. I might need to change the names / refactor it a little bit at some point. Let's not try to overthink it for now.

Reply:
@arhimondr : I am raising questions rather than proposing changes here. I agree we don't want to address them for **now** in the code.  But I usually find a bit thinking about the future direction useful.

Reply:
It makes sense. I aggree that `source` and `intermediate` don't work very well in case of a tasks that has both, inputs and table scans. Eventually these two methods might even be merged into a single one. Unfortunately It is very hard to say what is the right way before we try to prototype it =\

Reply:
@arhimondr : FWIW the code structure of `SectionExecutionFactory#createStageScheduler` might give some light on it. I start to smell convergence between RDD creation and stage scheduler creation... 😃